### PR TITLE
LienShark proposal for EigenApproWithGersgorin (second PR)

### PR DESCRIPTION
**Pull request by LienShark**
---
A good PR should include the following parts:

- [ ] Provide a clear title
- [ ] Provide meaningful descriptions in the PR and the commit logs
- [ ] Name your branch like ``<github-user-name>-hw#``, e.g., ``myname-hw2-attempt1``
- [ ] Merge from your specific branch (***not master***) to the `hw*` branch (***not master***, again)
- [ ] Exclude irrelevant files, such as binaries, text files, or dot files
- [ ] Make sure tests/CI pass

### CODE REVIEW

**Comment by LienShark** [comment]
I sincerely apologize for having to create a new pull request. I accidentally deleted the old remote branch, and since I am not entirely sure which branch matches it, I didn’t want to take any chances and risk causing further issues. I greatly appreciate everyone’s help!

In this submission, I’ve made the following changes:

1.I’ve added my entry to project/readme.rst.
2.I’ve provided further explanation on how I handled large matrices using incremental SVD and memory mapping to solve the problem.
3.I’ve clarified that the idea of working with million-row matrices was inspired by LLM models. This is a challenge I want to take on, though it’s just a goal. The fundamental issue is improving computation speed for large matrices.
4.I’ve added baseline comparisons, such as numpy.lineal.svd() and sklearn.decomposition.PCA().

File: project/LienShark/Readme.rst [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+My goal is to reduce memory and computing costs during data processing, giving the project practical application value.
+
+The baseline methods will include traditional SVD and spectral decomposition methods, which I aim to improve.
+
+
+######################
+System Architecture
+######################
+
+**1. Input and Output:**
```
Comment (yungyuc) [inline project/LienShark/Readme.rst:86]: I still did not see how you propose to deal with [array majoring](https://yyc.solvcon.net/en/latest/nsd/04matrix/matrix.html#row-major-2d-array).  I have been asking for this for [more than 3 weeks](https://github.com/yungyuc/nsdhw_24au/pull/47#pullrequestreview-2312311989).  Please specify how will you address it.  Please respond in a comment and update the proposal accordingly as soon as possible.

**Review by yungyuc — CHANGES_REQUESTED** [review]
- [x] [`L86`](https://github.com/yungyuc/nsdhw_24au/pull/94/files#diff-fec19e96fbfc351f95b3ac05432fbca60819a9075c25d6a8e59b4b783929c7aaR86): Specify how to deal with matrix majoring.

**Comment by yungyuc** [comment]
> I sincerely apologize for having to create a new pull request. I accidentally deleted the old remote branch, and since I am not entirely sure which branch matches it, I didn’t want to take any chances and risk causing further issues. I greatly appreciate everyone’s help!
> 
> In this submission, I’ve made the following changes:
> 
> 1.I’ve added my entry to project/readme.rst. 2.I’ve provided further explanation on how I handled large matrices using incremental SVD and memory mapping to solve the problem. 3.I’ve clarified that the idea of working with million-row matrices was inspired by LLM models. This is a challenge I want to take on, though it’s just a goal. The fundamental issue is improving computation speed for large matrices. 4.I’ve added baseline comparisons, such as numpy.lineal.svd() and sklearn.decomposition.PCA().

It's hard for me to see where you made the changes.  You should have added inline comments to highlight the points.

**Comment by yungyuc** [comment]
This PR succeeds #47.

File: project/README.rst [hunk 2]
```diff
@@ -82,6 +82,9 @@ Follow the format to add your project:
 14. `yang2829 <https://github.com/yang2829>`__ for
     `Voronoi Diagram <yang2829/README.rst>`__:
     https://github.com/yang2829/Voronoi-Diagram
+15. `LienShark <https://github.com/LienShark>`__ for
+    `EigenApproWithGersgorin <LienShark/README.rst>`__:
+    https://github.com/LienShark/EigenApproWithGersgorin
```
Comment (LienShark) [inline project/README.rst:87]: 1.added my entry to project/readme.rst.

File: project/LienShark/Readme.rst [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+**3. Module Division:**
+
+- Eigenvalue estimation module  
+- Dimensionality reduction module  
+- API module
+
+**4. Array Majoring:**
+
+- **Row-Major vs Column-Major:**
+  - **C++:**  I will use row-major order for better memory access.
```
Comment (ThreeMonth03) [inline project/LienShark/Readme.rst]: Are all C++ matrices in row-major order?

File: project/LienShark/Readme.rst [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+  - **Python (NumPy):** I will use column-major order for optimized operations.
```
Comment (ThreeMonth03) [inline project/LienShark/Readme.rst]: Why are the numpy matrices in column-major order instead of row-major order?

**Review by ThreeMonth03 — COMMENTED** [review]
I have some quick questions.

**Comment by yungyuc** [comment]
@LienShark are you ready for review?  You did not follow what I asked you to do: https://github.com/yungyuc/nsdhw_24au/pull/47#pullrequestreview-2312311989 and I am baffled.

**Comment by LienShark** [comment]
> @LienShark are you ready for review? You did not follow what I asked you to do: [#47 (review)](https://github.com/yungyuc/nsdhw_24au/pull/47#pullrequestreview-2312311989) and I am baffled.

sorry , I am trying to adjust my proposal and I will update  it as soon as possible

**Comment by LienShark** [comment]
I show why I want to use row-major and show some c++ code to explain how to do, thanks for everyone's help

File: project/LienShark/Readme.rst [hunk 5]
```diff
@@ -91,6 +91,55 @@ System Architecture
...
+
+    private:
+        vector<double> m_buffer;
+        int m_nrow;
+        int m_ncol;
+    };
+
 - **Output:**  
   1. The low-dimensional matrix reduced by PCA and SVD (non-approximate).
   2. Approximate eigenvalues estimated using the Gersgorin Theorem, and the low-dimensional matrix reduced by PCA.
```
Comment (ThreeMonth03) [inline project/LienShark/Readme.rst:145]: Is the output matrix in a row-majoring? Please write it to your proposal.

**Review by ThreeMonth03 — COMMENTED** [review]
- [ ] Write the method of the output matrix majoring.

**Comment by LienShark** [comment]
I show some c++ code to explain how to do in my output , thanks

File: project/LienShark/Readme.rst [hunk 6]
```diff
@@ -143,6 +143,42 @@ System Architecture
...
+  the Output will also row-major,because I want my term project has consistent
```
Comment (ThreeMonth03) [inline project/LienShark/Readme.rst]: Could you please proofread this sentence with ChatGPT?

**Review by ThreeMonth03 — COMMENTED** [review]
- [ ] Proofread the sentence.

**Comment by LienShark** [comment]
@yungyuc Did I meet the requirements? Or is there anything that still needs to be changed and strengthened?
@ThreeMonth03 I modified my sentence about the output

File: project/LienShark/Readme.rst [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+- **Input:**  
+  Massive matrices.  
+  For larger matrices, I plan to experiment with either Incremental SVD or memory mapping to address memory limitations.  
+  I will start with smaller matrices to ensure feasibility but aim to work with million-row matrices eventually.  
+  If that proves impractical, I will adjust my approach accordingly.
+
+  I will adopt the row-major approach to store my matrix because, in most cases, a row in a matrix typically represents a new data sample,
+  while a column represents its features. When performing Incremental SVD, I tend to add more rows incrementally, as this approach is more meaningful in practice. 
+  It also helps reduce computational complexity and simplifies storage management.
+  For e.g: Today, if I need to process a matrix larger than 1000x1000, I would first load a portion of it, and then incrementally update the SVD one row at a time
```
Comment (yungyuc) [inline project/LienShark/Readme.rst:97]: The progressive reading (updating rows/columns one by one) works for both row- and column-majoring.  Do not hold the wrong idea that row-majoring is better for the progressive reading.

**Review by yungyuc — APPROVED** [review]
It's OK.  But be reminded, @LienShark, you did not know very well about the problem you propose to deal with.  You may have hard time writing code and making slides.

**Pull request merged**